Bump snaps packages and support dynamic interfaces

FrederikBolding opened the pull request:
## **Description**
Bump snaps packages to latest and deals with breaking changes.

Summary of changes in the snaps deps:

- **BREAKING:** Update permission format for name lookup Snaps
- Add dynamic interfaces
  -  Can be created with `snap_createInterface` and updated with `snap_updateInterface`
  - All existing UI is backwards compatible
- Add support for Snap defined timeouts
  - A Snap can now define `maxRequestTime` to extend or shorten its execution timeout
- Loosen allowlist requirements
  - Snaps with certain permissions can now be installed without being allowlisted
- Reintroduced `DecompressionStream` for improved installation performance
- Endowment permission specifications were moved requiring imports to be updated
- Bump tar-stream letting us remove a patch

This PR also contains https://github.com/MetaMask/metamask-extension/pull/22795 merged in, which contains the integration for dynamic interfaces.

## Related issues
Closes https://github.com/MetaMask/MetaMask-planning/issues/1954

## **Manual testing steps**

1. Fire up the interactive UI test snap https://metamask.github.io/snaps/test-snaps/latest/
2. Make sure it works as intended

FrederikBolding committed:
Bump snaps packages

FrederikBolding committed:
Update LavaMoat policies

FrederikBolding committed:
Fix imports

FrederikBolding committed:
Bump test-snaps

FrederikBolding committed:
Fix lint and some tests

FrederikBolding committed:
Fix another import

FrederikBolding committed:
Fix Socket warning

FrederikBolding committed:
Fix name lookup

FrederikBolding committed:
Fix tests

FrederikBolding committed:
Use latest snaps-sdk everywhere

hmalik88 commented on app/scripts/lib/SnapsNameProvider.ts:
```diff
@@ -134,9 +134,12 @@ export class SnapsNameProvider implements NameProvider {
 
-      proposedNames = domain ? [domain] : [];
+      // TODO: Determine if this is what we want.
+      proposedNames = domains
+        ? domains.map((domain) => domain.resolvedDomain)
+        : [];
```
```suggestion
      proposedNames = domains
        ? [...new Set(domains.map((domain) => domain.resolvedDomain))]
        : [];
```

The snap can potentially (erroneously) return dupes, we should guard against that.

FrederikBolding committed:
Update app/scripts/lib/SnapsNameProvider.ts

Co-authored-by: Hassan Malik <41640681+hmalik88@users.noreply.github.com>

FrederikBolding committed:
Fix import after rebase

GuillaumeRx committed:
Interactive Interfaces (#22795)

## **Description**

This PR integrates the necessary changes for interactive Interfaces.

- Setup and update the various controllers.
- Update the `SnapUIRenderer` to handle an `interfaceId` rather than
direct content.
- Add the logic for interface state and event handling in a new
`InterfaceContext` react context.
- Update the different APIs to handle the newly returned values.
- Move the UI components mapping out of the `SnapUIRenderer` file
- Rework confirmations to take the `SnapUIRenderer`.

## **Related issues**

## **Manual testing steps**

1. Fire up the interactive UI test snap [coming soon]
2. Play with the snap

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: Frederik Bolding <frederik.bolding@gmail.com>

GuillaumeRx committed:
fix insights warning UI

FrederikBolding committed:
Bump deps again

FrederikBolding committed:
Update yarn.lock

FrederikBolding committed:
Fix lint

FrederikBolding committed:
Fix lint and E2E

FrederikBolding committed:
Fix type issues

FrederikBolding committed:
Delete broken story

FrederikBolding commented on ui/components/app/metamask-translation/metamask-translation.stories.js:
With the blessing of @brad-decker, I've deleted this story as it is broken due to circular dependencies.

FrederikBolding committed:
Bump snaps-controllers again

FrederikBolding commented:
@SocketSecurity ignore npm/@metamask/snaps-controllers@5.0.1
just published.

Mrtenz commented on app/scripts/lib/SnapsNameProvider.ts:
```diff
@@ -135,9 +135,12 @@ export class SnapsNameProvider implements NameProvider {
+      const domains = result?.resolvedDomains;
 
-      proposedNames = domain ? [domain] : [];
+      // TODO: Determine if this is what we want.
```
What do we need to determine here?

FrederikBolding replied:
We need to determine how we are gonna map from the new return format to the format that the Petnames system expects.

@hmalik88 is gonna be working on that after this PR is merged, just left it as a note: https://github.com/MetaMask/MetaMask-planning/issues/1959

Mrtenz commented on package.json:
```diff
@@ -218,7 +218,7 @@
     "@trezor/connect-web": "9.0.11",
     "lavamoat-core@npm:^15.1.1": "patch:lavamoat-core@npm%3A15.1.1#~/.yarn/patches/lavamoat-core-npm-15.1.1-51fbe39988.patch",
-    "tar-stream@npm:^3.1.6": "patch:tar-stream@npm%3A3.1.6#~/.yarn/patches/tar-stream-npm-3.1.6-ce3ac17e49.patch"
+    "@metamask/snaps-sdk": "^2.1.0"
```
Why is this resolution necessary?

FrederikBolding replied:
This is for typing reasons. If we don't use a resolution we get clashes between the multiple instances of `SnapId`.

Mrtenz commented on ui/components/app/snaps/snap-insight/snap-insight.js:
```diff
@@ -68,8 +64,14 @@ export const SnapInsight = ({
+    return () => {
+      interfaceId && dispatch(deleteInterface(interfaceId));
+    };
+  }, []);
```
```suggestion
  }, [interfaceId]);
```

david0xd commented on ui/components/app/snaps/snap-ui-renderer/components/form.ts:
```diff
@@ -0,0 +1,18 @@
+export const form: UIComponentFactory<Form> = ({ element, ...params }) => ({
+  element: 'SnapUIForm',
+  // @ts-expect-error This is a problem with the types generated in the snaps repo.
+  children: element.children.map((children) =>
```
If this is a typing issue that is not easy to solve, what are the arguments against using `(children: object)` or `(children: any)` and using `// @ts-expect-error` instead?

To me it looks like the `// @ts-expect-error` should be used only when there are no other solutions, it's kinda the last thing to do. I acknowledge that using `any` or `object` is bad practice, but `// @ts-expect-error` also looks even more bad IMO. WDYT?

FrederikBolding replied:
I was thinking we remove these once the TypeScript problems are solved in the snaps repo instead. If there is no comment it feels like more likely that we forget about it?

GuillaumeRx replied:
Yes I would prefer leaving a `ts-expect-error` for things we want to resolve later

david0xd replied:
```
If there is no comment it feels like more likely that we forget about it?
```
It can also be ``// TODO: ...```. But, if you prefer keeping `@ts-expect-error` temporarily, I'm alright with that.

david0xd replied:
I will mark this as resolved. Answered here what's my opinion on such stuff: https://github.com/MetaMask/metamask-extension/pull/22828#discussion_r1496308545

GuillaumeRx committed:
add missing deps to various useEffect and fix wrong prop type

david0xd commented on ui/components/app/snaps/snap-ui-renderer/utils.ts:
```diff
@@ -0,0 +1,17 @@
+  params.elementKeyIndex += 1;
+  const indexKey = `snap_ui_element_${type}__${params.elementKeyIndex}`;
+  // @ts-expect-error This is a problem with the types generated in the snaps repo.
+  const mapped = COMPONENT_MAPPING[type](params as any);
```
I would like to avoid having `@ts-expect-error` here if the TypeScript is not able to Infer the key type properly. It is a string anyway at the end.
So, adding helper/utility type to handle this would help removing `@ts-expect-error`.

This can be the helper type:
```typescript
export type ComponentType =
  | 'panel'
  | 'heading'
  | 'text'
  | 'divider'
  | 'spinner'
  | 'image'
  | 'copyable'
  | 'row'
  | 'address'
  | 'button'
  | 'form'
  | 'input';
```
And the problematic line can be fixed with:
```typescript
const mapped = COMPONENT_MAPPING[type as ComponentType](params as any);
```
`ComponentType` can probably have a better name, anyway.

FrederikBolding replied:
I would rather not introduce this type even temporarily because we are trying not to duplicate code into the client repos. These types worked before the extension bumped the TypeScript version. 

The type export problems should be fixed when we ship https://github.com/MetaMask/snaps/pull/2120 AFAIK, so it seems fine to wait for that IMO.

david0xd replied:
Alright. But, I would suggest that in the future we maybe wait for such stuff to get resolved/fixed/merged, then do a release. It feels weird to merge stuff with TypeScript errors expected and error ignore patterns that are in functional code, I'm fine having them only in the test files.
Anyway, I understand the issues here, so I'm not blocking PR, I can approve without fix for this.

FrederikBolding replied:
Agreed, but this problem started occurring while this PR was in development and not before (due to TypeScript being updated in the extension I believe). If it had been a problem from the start I would be fully in agreement.

david0xd replied:
Sure, no problem. Let's fix it in the next release.

david0xd commented on ui/selectors/selectors.js:
```diff
@@ -894,6 +894,63 @@ export const getAnySnapUpdateAvailable = createSelector(
+  (snapInterface) => snapInterface,
+);
+
+export const getInterfaceContent = createSelector(
```
To stay consistent with our coding standards, this should have docstring.

FrederikBolding replied:
I don't find docstrings that useful for selectors since they rely on global state more than inputs 🤔

david0xd replied:
I still find it useful. Especially after a while, when coming back to the certain parts of the code we merged. It can be easier to understand some functionality and purpose of the code, find and navigate through the code or find related functionality elements involved in the same feature.

I know that everybody have their own approach to this. Sometimes we just add a docstring to fullfil the "consistency requirement", but sometimes we add it to explain something and make it easier to understand for everybody. I think we should do it for both reasons.

For example this is very nice and helpful:
![Screenshot 2024-02-20 at 19 35 07](https://github.com/MetaMask/metamask-extension/assets/13301024/19a4f933-dd51-4460-86ed-d775502f3397)

On the other side this looks a bit confusing or even unrelated to Snaps features:
![Screenshot 2024-02-20 at 19 37 01](https://github.com/MetaMask/metamask-extension/assets/13301024/eaadb878-1c3e-4f4a-bea9-510b16a469ff)

Of course I can do a full-text search or find usages, use the other tools, etc. But, when looking at the code and trying to understand something, good docstring can be very helpful. 

Good docstring can explain the purpose, explain the functionality more, leave some space for additional explanation in the future when code is changed, gives the function more integrity from a developer's perspective, etc.

Chances are that we'll onboard new developers who will work on this and might need better/easier understanding and navigation through the code. Also, chances are that somebody else will add bunch of code in between this and make understanding and finding stuff more difficult (this file is huge). Then, chances are that this file will be refactored in the future and these functions split, etc.

These are probably the major reasons why I like to see docstrings added. Minor thing, but still nice to have and can be helpful.

I can approve with or without docstrings added in this case, this is not a blocker.

FrederikBolding replied:
Fair point! @GuillaumeRx do you want to add some docstrings here before we get this merged?

GuillaumeRx replied:
Yes I can add some docstrings here

FrederikBolding commented on ui/components/app/snaps/snap-ui-renderer/components/panel.ts:
```diff
@@ -0,0 +1,21 @@
+import { mapToTemplate } from '../utils';
+import { UIComponentFactory } from './types';
+
+export const panel: UIComponentFactory<Panel> = ({ element, ...params }) => ({
```
Hmm, we haven't in the past written docstrings for components etc in the extension. We could start doing it, but I don't see a lot of added value for it right now tbh

david0xd replied:
My answer would be similar to this one here: https://github.com/MetaMask/metamask-extension/pull/22828#discussion_r1496337308

But, in this case, the integrity of the components is being held by the Snaps SDK where all of these have docstring. So, nice to have, but it's fine either way. _Still, being consistent propagates good coding practices from our side_ :).

david0xd commented on ui/selectors/selectors.js:
```diff
@@ -894,6 +894,63 @@ export const getAnySnapUpdateAvailable = createSelector(
   },
 );
+
+export const getMemoizedTargetSubjectMetadata = createDeepEqualSelector(
```
Resolving following the conversation here: https://github.com/MetaMask/metamask-extension/pull/22828#discussion_r1495878427

GuillaumeRx committed:
add missing doc stings to selectors

GuillaumeRx committed:
fix typo

FrederikBolding commented on ui/selectors/selectors.js:
```diff
@@ -894,6 +894,72 @@ export const getAnySnapUpdateAvailable = createSelector(
+);
+
+/**
+ * Get a memoized verison of the content from a Snap interface with a given ID.
```
```suggestion
 * Get a memoized version of the content from a Snap interface with a given ID.
```

FrederikBolding committed:
Update ui/selectors/selectors.js
